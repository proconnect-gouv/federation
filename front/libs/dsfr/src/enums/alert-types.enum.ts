/* istanbul ignore file */
export enum AlertTypes {
  ERROR = 'error',
  INFO = 'info',
  SUCCESS = 'success',
  WARNING = 'warning',
}
