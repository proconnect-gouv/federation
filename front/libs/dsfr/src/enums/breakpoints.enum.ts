/* istanbul ignore file */
export enum Breakpoints {
  XS = 576,
  SM = 768,
  MD = 992,
  LG = 1248,
}
