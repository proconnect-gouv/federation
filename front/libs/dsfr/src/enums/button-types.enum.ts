/* istanbul ignore file */
export enum ButtonTypes {
  SUBMIT = 'submit',
  BUTTON = 'button',
  RESET = 'reset',
}
