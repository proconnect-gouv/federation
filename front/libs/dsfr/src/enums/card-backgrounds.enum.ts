export enum CardBackgrounds {
  GREY = 'grey',
  NO_BACKGROUND = 'no-background',
  NO_BORDER = 'no-border',
  SHADOW = 'shadow',
}
