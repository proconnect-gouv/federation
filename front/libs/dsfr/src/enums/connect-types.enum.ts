export enum ConnectTypes {
  FRANCE_CONNECT = 'FranceConnect',
  PRO_CONNECT = 'ProConnect',
}
