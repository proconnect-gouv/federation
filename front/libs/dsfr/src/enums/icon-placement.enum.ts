/* istanbul ignore file */

// declarative file
export enum IconPlacement {
  LEFT = 'left',
  RIGHT = 'right',
}
