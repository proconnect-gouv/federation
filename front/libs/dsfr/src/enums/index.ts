/* istanbul ignore file */

// declarative file
export * from './alert-types.enum';
export * from './button-types.enum';
export * from './icon-placement.enum';
export * from './pagination.enum';
export * from './priorities.enum';
export * from './sizes.enum';
