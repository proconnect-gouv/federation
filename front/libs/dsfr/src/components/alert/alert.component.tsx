import classnames from 'classnames';
import type { PropsWithChildren } from 'react';
import React from 'react';

import { EventTypes, HeadingTag, type PropsWithClassName } from '@fc/common';
import { t } from '@fc/i18n';

import { Sizes } from '../../enums';

interface AlertComponentV2Props extends PropsWithClassName, PropsWithChildren {
  // @NOTE [DSFR] attribute role="alert"
  // - should be defined if the Component is injected dynamicly into the page
  // - should NOT be defined if the Component is not injected dynamicly into the page
  noRole?: boolean;
  size?: Omit<Sizes, Sizes.LARGE>;
  type?: EventTypes;
  title?: string;
  // @NOTE TS Omit(H1) issued with HeadingTag's type
  heading?: HeadingTag.H2 | HeadingTag.H3 | HeadingTag.H4 | HeadingTag.H5 | HeadingTag.H6;
  onClose?: () => void;
  dataTestId?: string;
}

export const AlertComponent = React.memo(
  ({
    children: description,
    className,
    dataTestId = 'AlertComponent',
    heading: Heading = HeadingTag.H3,
    noRole = false,
    onClose = undefined,
    size = Sizes.MEDIUM,
    title = undefined,
    type = EventTypes.INFO,
  }: AlertComponentV2Props) => {
    const showTitle = size !== Sizes.SMALL && !!title;
    const closeLabel = (onClose && t('DSFR.alert.close')) || undefined;

    return (
      <div
        className={classnames(className, `fr-alert fr-alert--${type} fr-alert--${size}`)}
        data-testid={dataTestId}
        role={noRole ? undefined : 'alert'}>
        {showTitle && (
          <Heading className="fr-alert__title" data-testid={`${dataTestId}-title`}>
            {title}
          </Heading>
        )}
        {description}
        {!!onClose && (
          <button
            className="fr-link--close fr-link"
            data-testid={`${dataTestId}-close-button`}
            title={closeLabel}
            onClick={onClose}>
            {closeLabel}
          </button>
        )}
      </div>
    );
  },
);

AlertComponent.displayName = 'AlertComponent';
