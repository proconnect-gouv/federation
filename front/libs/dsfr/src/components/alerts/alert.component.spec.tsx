import { render } from '@testing-library/react';

import { AlertTypes, Sizes } from '../../enums';
import { AlertComponent } from './alert.component';

describe('Alert', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot with default values', () => {
    // when
    const { container } = render(<AlertComponent>Children</AlertComponent>);

    // then
    expect(container).toMatchSnapshot();
  });

  it('should remove the role="alert" attribute if noRole is true', () => {
    // when
    const { getByTestId } = render(<AlertComponent noRole>Children</AlertComponent>);
    const element = getByTestId('AlertComponent');

    // then
    expect(element).not.toHaveAttribute('role');
  });

  it('should have class fr-alert--md by default when size is not defined', () => {
    // when
    const { getByTestId } = render(<AlertComponent>Children</AlertComponent>);
    const element = getByTestId('AlertComponent');

    // then
    expect(element).toHaveClass('fr-alert--md');
  });

  it('should have class fr-alert--info by default when type is not defined', () => {
    // when
    const { getByTestId } = render(<AlertComponent>Children</AlertComponent>);
    const element = getByTestId('AlertComponent');

    // then
    expect(element).toHaveClass('fr-alert--info');
  });

  it('should have class fr-alert--sm if size is "sm"', () => {
    // when
    const { getByTestId } = render(<AlertComponent size={Sizes.SMALL}>Children</AlertComponent>);
    const element = getByTestId('AlertComponent');

    // then
    expect(element).toHaveClass('fr-alert--sm');
  });

  it('should have class fr-alert--error if type is ERROR', () => {
    // when
    const { getByTestId } = render(
      <AlertComponent type={AlertTypes.ERROR}>Children</AlertComponent>,
    );
    const element = getByTestId('AlertComponent');

    // then
    expect(element).toHaveClass('fr-alert--error');
  });

  it('should have class fr-alert--info if type is INFO', () => {
    // when
    const { getByTestId } = render(
      <AlertComponent type={AlertTypes.INFO}>Children</AlertComponent>,
    );
    const element = getByTestId('AlertComponent');

    // then
    expect(element).toHaveClass('fr-alert--info');
  });

  it('should have class fr-alert--success if type is SUCCESS', () => {
    // when
    const { getByTestId } = render(
      <AlertComponent type={AlertTypes.SUCCESS}>Children</AlertComponent>,
    );
    const element = getByTestId('AlertComponent');

    // then
    expect(element).toHaveClass('fr-alert--success');
  });

  it('should have class fr-alert--warning if type is WARNING', () => {
    // when
    const { getByTestId } = render(
      <AlertComponent type={AlertTypes.WARNING}>Children</AlertComponent>,
    );
    const element = getByTestId('AlertComponent');

    // then
    expect(element).toHaveClass('fr-alert--warning');
  });

  it('should have class injected by className props', () => {
    // when
    const { getByTestId } = render(<AlertComponent className="foo bar">Children</AlertComponent>);
    const element = getByTestId('AlertComponent');
    // then
    expect(element).toHaveClass('foo bar');
  });

  it('should have a custom testid', () => {
    // when
    const { getByTestId } = render(
      <AlertComponent dataTestId="alert-component-custom-testid" type={AlertTypes.WARNING}>
        Children
      </AlertComponent>,
    );
    const element = getByTestId('alert-component-custom-testid');

    // then
    expect(element).toBeInTheDocument();
  });
});
