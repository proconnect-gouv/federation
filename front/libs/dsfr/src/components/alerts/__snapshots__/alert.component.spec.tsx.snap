// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert should match the snapshot with size medium and type error 1`] = `
<div>
  <div
    class="fr-alert fr-alert--error"
    data-testid="AlertComponent"
  >
    Children
  </div>
</div>
`;
